UITests/Tests/AxeScanAllTests.cs

```diff
@@ -0,0 +1,124 @@
+using System.IO;
+using System.Collections.ObjectModel;
+
+namespace UITests.Tests
+{
+    [TestClass]
+    public class AxeScanAll : TestBase
+    {       
+        public static readonly string jsonUri = "ms-appx:///../../../../../../WinUIGallery/DataModel/ControlInfoData.json";
+        public static new WindowsDriver<WindowsElement> Session => SessionManager.Session;
+
+        public class ControlInfoData
+        {
+            public ObservableCollection<ControlInfoDataGroup> Groups { get; set; }
+        }
+
+        public class ControlInfoDataGroup
+        {
+            public string UniqueId { get; set; }
+            public string Title { get; set; }
+            public string Subtitle { get; set; }
+            public string Description { get; set; }
+            public string ImagePath { get; set; }
+            public string ImageIconPath { get; set; }
+            public string ApiNamespace { get; set; }
+            public bool IsSpecialSection { get; set; }
+            public string Folder { get; set; }
+            public ObservableCollection<ControlInfoDataItem> Items { get; set; }
+        }
+
+        public class ControlInfoDataItem
+        {
+            public string UniqueId { get; set; }
+            public string Title { get; set; }
+            public string ApiNamespace { get; set; }
+            public string Subtitle { get; set; }
+            public string Description { get; set; }
+            public string ImagePath { get; set; }
+            public string ImageIconPath { get; set; }
+            public string BadgeString { get; set; }
+            public string Content { get; set; }
+            public bool IsNew { get; set; }
+            public bool IsUpdated { get; set; }
+            public bool IsPreview { get; set; }
+            public bool HideSourceCodeAndRelatedControls { get; set; }
+            public ObservableCollection<ControlInfoDocLink> Docs { get; set; }
+            public ObservableCollection<string> RelatedControls { get; set; }
+
+            public bool IncludedInBuild { get; set; }
+        }
+
+        public class ControlInfoDocLink
+        {
+            public ControlInfoDocLink(string title, string uri)
+            {
+                this.Title = title;
+                this.Uri = uri;
+            }
+            public string Title { get; set; }
+            public string Uri { get; set; }
+        }
+
+        private static ControlInfoData controlInfoData;
+
+        private string[] ExclusionList =
+        {
+            "WebView2" // 46668961: Web contents from WebView2 are throwing null BoundingRectangle errors.
+        };
+
+        [ClassInitialize]
+        public static void ClassInitializeAsync(TestContext context)
+        {
+            ParseJson();
+        }
+
+        [TestMethod]
+        [TestProperty("Description", "Scan all controls in the WinUIGallery for accessibility issues.")]
+        public void ValidateAccessibilityWithAxe()
```

Comment:
This will result in a single test case that covers every page in the app. This has some advantages. One is that it takes a very long time to run. Another is that if it fails on one page, it will stop on failure and the rest of the pages will not execute. And if you want to re-run the test to investigate a failure you have to run the whole thing, even if you only care about a single page.

A better approach would be to use data-driven testing. With this approach you have a single test method, but it is run multiple times with different data each time.

This is actually how we test WinUI Gallery in the WinUI3 repo. Take a look at:
controls\test\MUXControls.Test\ScenarioAppTests\WinUIGalleryDesktopTests.cs

The test method has this attribute:
```
[DataSource("Table:WinUIGalleryTestData.xml#MenusAndToolbars")]
```
Which means that it will pull in data from the file "WinUIGalleryTestData.xml". This is a file that is generated at build time by parsing ControlInfoData.json.

You won't be able to directly re-use this code. The WinUI3 repo uses TAEF, but this repo uses MSTest. But MSTest also supports data-driven testing.

I haven't uses MSTest's data driven features before, so I can't comment on exactly how to use it. But there is documentation here:
[https://learn.microsoft.com/en-us/visualstudio/test/how-to-create-a-data-driven-unit-test?view=vs-2022](https://learn.microsoft.com/en-us/visualstudio/test/how-to-create-a-data-driven-unit-test?view=vs-2022)

These docs seem to imply that only a sql db can be used as a source. But it sounds like it should be possible to use xml and other formats. Have a look around and see what you can find. I came across these articles:
[https://www.meziantou.net/mstest-v2-data-tests.htm](https://www.meziantou.net/mstest-v2-data-tests.htm)
[https://visualstudiomagazine.com/articles/2009/09/15/unit-testing-with-vsts2008-part-3.aspx](https://visualstudiomagazine.com/articles/2009/09/15/unit-testing-with-vsts2008-part-3.aspx)
[https://mariusbancila.ro/blog/2019/11/28/data-driven-unit-tests-for-managed-code/](https://mariusbancila.ro/blog/2019/11/28/data-driven-unit-tests-for-managed-code/)

Reply:
Thanks for the info. I've updated the tests do be data driven testing. Since `TestContext.DataRow` isn't supported in .net, I'm using the DynamicData attribute as recommended in the docs.